Convert workers configuration form to DDF

rvsia opened the pull request:
fixes https://bugzilla.redhat.com/show_bug.cgi?id=1672437

**Description**

- Converts old form to new schema
- Allows to show custom values in selects

Data consistency:

- All values are converted into bytes in the form
- All values are shown in human size in the select labels
- All values are converted into ruby methods format after submitting

TODO:

- [x] submit
- [x] tests

**Before** 

![image](https://user-images.githubusercontent.com/32869456/57625031-5f3d3580-7593-11e9-8042-a912e5cda2bc.png)

**After**

![image](https://user-images.githubusercontent.com/32869456/57625039-6401e980-7593-11e9-99d4-aa1023de77bd.png)


---

@Fryguy 
In the BZ there is said: '_Some dropdowns are not big enough._'

I have no idea which ones and how big they should be. :smile: The best solution will be to come up with one range which could be used by all selects. 

@kbrock I heard you could help.

terezanovotna commented:
This looks good @rvsia !
When the the text is too long, I recommend aggregating it and then showing the full length on hover

rvsia commented:
@terezanovotna Which text do you mean?

rvsia commented:
@miq-bot add_label hammer/no, react, refactoring, enhancement, settings

terezanovotna commented:
in the dropdown

rvsia commented:
@miq-bot add_reviewer @Hyperkid123 @terezanovotna 

@terezanovotna I added a helper text with warning error below UI Worker select. Originally on Master, it should show a warning message with that text but it doesn't work for me (probably a :bug: ). However, what do you think about showing the message as a helperText instead of a message? With this solution, a user see the message the whole time and doesn't need to look up.

rvsia commented:
@miq-bot add_reviewer @Hyperkid123 
@miq-bot add_reviewer @terezanovotna

Hyperkid123 commented on app/javascript/components/workers-form/helpers.jsx:
```diff
@@ -0,0 +1,192 @@
+/* Transform input value to bytes (12545411) */
+export const toBytes = (value) => {
```
js doc would be nice for functions like these. I think it will be reused in more places in future.

Hyperkid123 commented on app/javascript/components/workers-form/workers-form.jsx:
```diff
@@ -0,0 +1,178 @@
+/* global add_flash */
```
Can you add this to .eslintrc? I think we can add this global everywhere

Hyperkid123 commented on app/javascript/components/workers-form/workers-form.jsx:
```diff
@@ -0,0 +1,178 @@
+} from './helpers';
+
+const WorkersForm = ({ server, product, zone }) => {
+  const { id, name } = server;
```
You can use spread operator in the params

Hyperkid123 commented on app/javascript/components/workers-form/workers-form.jsx:
```diff
@@ -0,0 +1,178 @@
+    return API.patch(`/api/servers/${id}/settings`, patch)
+      .then(
+        () => {
+          add_flash(__(`Configuration settings saved for ${product} Server "${name} [${id}]" in Zone "${zone}"`), 'success');
```
Translation wont work you need to use sprintf:
```
sprintf(__('%s:"%s" was successfully deleted'), toolbar.entity, entityName)
```

GetText does not get template strings

Hyperkid123 commented on app/javascript/components/workers-form/workers.schema.js:
```diff
@@ -0,0 +1,233 @@
+} from './helpers';
+
+function addSchema(formValues) {
+  if (!formValues) return null;
```
I would throw and Error if you don't pass the form values param

Hyperkid123 commented on app/javascript/components/workers-form/workers.schema.js:
```diff
@@ -0,0 +1,233 @@
+        title: __('Generic Workers'),
+        fields: [
+          {
+            component: componentTypes.SELECT_COMPONENT,
```
use just `SELECT`

Hyperkid123 commented on app/javascript/components/dual-group/index.jsx:
```diff
@@ -0,0 +1,14 @@
+const Dualgroup = ({ fields, formOptions }) => (
+  <React.Fragment>
+    <Row>
+      <Col md={6}>{formOptions.renderForm([fields[0]], formOptions)}</Col>
```
I would probably use something like leftColumn/rightColumn or column size based on field count so we can have 
1, 2, 3, 4, 6, 8, 10 and 12 columns.

rvsia commented:
@Hyperkid123 Fixed all your notes! :+1:

Fryguy commented:
cc @agrare

agrare commented on app/javascript/components/workers-form/helpers.jsx:
```diff
@@ -0,0 +1,237 @@
+  if (!value) return undefined;
+
+  if (value >= 1073741824) {
+    return `${value / (1024 ** 3)} GB`;
```
Not sure how OCD we want to be but base 2 is GiB/MiB/KiB (https://en.m.wikipedia.org/wiki/Gibibyte)

rvsia replied:
I know, but this is how it is in the current form. Value is in GiB, etc. but label shows GB. 😖 I am not sure if we want to change it, or why it is made this way. @h-kataria

agrare replied:
:+1: makes sense to keep it the same as it is now

himdel replied:
We already had all of those helpers in `miq_formatters.js`, to match the rails formatters for charts.

We should probably merge this with those (this is `number_to_human_size`).

(`toRubyMethod` being an exception)

Fryguy commented on app/javascript/components/workers-form/helpers.jsx:
```diff
@@ -0,0 +1,237 @@
+    if (value.match(/kilobytes/)) {
+      return sizeValue * 1024;
+    }
+    if (value.match(/.bytes/)) {
```
This line has `.bytes`, but the others don't have the leading dot.  Is that intentional?  Also, a dot in Regex means "any character"...what that also intentional or did you mean `\.`?

rvsia replied:
You are right! It should be \.

Fryguy commented on app/javascript/components/workers-form/helpers.jsx:
```diff
@@ -0,0 +1,237 @@
+  }
+  return value
+    .replace('.megabytes', ' MB')
+    .replace('.gigabytes', ' GB')
```
Minor, but this list should be sorted by reverse size... GB, MB, KB, B

Fryguy commented on app/javascript/components/workers-form/helpers.jsx:
```diff
@@ -0,0 +1,237 @@
+/**
```
I feel like we shouldn't be writing (most of) this ourselves.  There has got to be a package that does presentation of bytes.  A quick google showed me https://www.npmjs.com/package/pretty-bytes, but maybe lodash has something?

rvsia replied:
I understand, there are a lot of options, but I didn't want to add another dependency for such simple function.

h-kataria replied:
@rvsia i agree if we need to add similar calculations/functionality on any other screen we can look for a package in future but for this PR we can go with what you have now.

h-kataria commented:
@rvsia tested in UI, seems to be working fine. Are all the issues/comments addressed, ready to be merged?

h-kataria commented:
@rvsia after some more testing found couple of issues:
- after i press Refresh button on Workers tab, i see double set of form buttons on the screen.
![double_form_buttons](https://user-images.githubusercontent.com/3450808/58587886-6bc8bb80-822c-11e9-92d8-ad2cd1f282b4.png)

- I tried to save a custom value on Advanced settings tab for Generic worker, then i went back to Workers tab and changed the value to 200 MB and pressed Save. I see  aflash message saying that the value was saved, and all looks good, but when i go away from that screen and come back, i see the custom value back again in Generic Workers drop down, so it looks like the custom value is not being updated to the one i had selected before i pressed save button.

![Screenshot from 2019-05-29 16-13-46](https://user-images.githubusercontent.com/3450808/58588040-c235fa00-822c-11e9-9958-16c0acb58562.png)

rvsia commented:
@h-kataria thank you for noticing it!

1) Buttons was handled in replace_right_cell, however after refreshing the page only render is called, so I added another condition which checks if the current tab is workers settings and sets `x_edit_buttons_locals` to nil.

2) There was a typo: instead of `.` was `_` in the name of the variable.

Both issues should be fixed now.

h-kataria commented:
Verified fixes in UI, seems to be working great. Ready to be merged once tests are green.

h-kataria commented:
@rvsia can you take a look at spec test faiure

rvsia commented:
@h-kataria spec caused by https://github.com/ManageIQ/manageiq-ui-classic/pull/5609 -> fixed (the spec is obsolete)

However, https://github.com/ManageIQ/manageiq-ui-classic/pull/5649 and https://github.com/ManageIQ/manageiq-ui-classic/pull/5650 introduced new styling for React forms and this form now looks different:

![image](https://user-images.githubusercontent.com/32869456/58783705-d9d10200-85e1-11e9-8785-530efee69012.png)

There is a issue with responsiveness:  on smaller screens dropdowns are overlapping each other:


![image](https://user-images.githubusercontent.com/32869456/58783789-13a20880-85e2-11e9-9923-f739b3bd9e9f.png)

Dropdowns have min-width set bigger than their layout space. 

So, I would like to ask @epwinchell @terezanovotna @Hyperkid123 Should this form look same as before (that means it should have 100 % width)? Then, I would try to fix this issue in this PR. If not, we need to change this min-width in data-driven-forms library (and I don't know if this PR should wait for it).

Hyperkid123 commented:
@rvsia i think we should use the previous version. Right now the new form wrapper is not optional. If you can you can modify the renderer wrapper (where we set up the layout mapper) and add a prop by which we can switch the form wrapper layout between the new one and the old one.

rvsia commented:
@Hyperkid123 what about to be able to rewrite the classname of the form wrapper?

epwinchell commented:
@rvsia Removing the second column will eliminate the issue.

Hyperkid123 commented:
@rvsia sure that si one way to do it.

@epwinchell i don't think there are any plans for multi column forms. @terezanovotna what do you think about having just one column?

epwinchell commented:
@miq-bot add_reviewer @epwinchell

epwinchell commented:
@miq-bot add-reviewer @epwinchell

epwinchell commented:
@miq-bot add_reviewer @epwinchell

rvsia commented:
@Hyperkid123 I updated FormRenderer:  it passes a className to `<Form>` and default is set to `form-react`, so anyone can overwrite it. I also deleted FormWrapper because of passing className prop. Was there any reason why it was in a separate component `<Form {...props}>{Children}..`? It looks that's working well without it.

Hyperkid123 commented:
@rvsia i didn't want it to be anonymous function so we can actually se the component inside the Dev tools in virtual DOM so we can debug it if need be.

epwinchell commented:
@rvsia Please replace screenshots when you convert to single column. Thanks.

rvsia commented:
@epwinchell After changes in form-renderer (it allows to overwrite the className of the form) the form looks same as in the PR description.

martinpovolny commented:
@rvsia : Please, rebase.

@epwinchell : can this one be merged?

rvsia commented:
@martinpovolny tests are failling after rebas, I will look on them when i have some free time 💭

rvsia commented:
@martinpovolny @Hyperkid123 I discovered why tests failed. In Orchestration template fetchMock endpoints were defined after mounting the component (and it looks like the older version of enzyme works different than newer one - the newer one probably know how to work with hook components better, with the old version my tests don't work...)

Right now it is all green! :green_book:
